utils/frame/try-runtime/remote-externalities/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+//! through which you can infer storage types. There are two options here:
+//!
+//! 1. Build a mock runtime, similar how to you would build one in a pallet test (see example
+//!    below). The very important point here is that this mock needs to hold real values for types
+//!    that matter for you, based on the chain of interest. Some typical ones are:
+//!
+//! - `sp_runtime::AccountId32` as `AccountId`.
+//! - `u32` as `BlockNumber`.
+//! - `u128` as Balance.
+//!
+//! Once you have your `Runtime`, you can use it for storage type resolution and do things like
+//! `<my_pallet::Pallet<Runtime>>::storage_getter()` or `<my_pallet::StorageItem<Runtime>>::get()`.
+//!
+//! 2. Or, you can use a real runtime.
+//!
+//! ### Example
+//!
+//! With a test runtime
+//!
+//! ```ignore
+//! use remote_externalities::Builder;
+//!
+//! #[derive(Clone, Eq, PartialEq, Debug, Default)]
+//! pub struct TestRuntime;
+//!
+//! use frame_system as system;
+//! impl_outer_origin! {
+//!     pub enum Origin for TestRuntime {}
+//! }
+//!
+//! impl frame_system::Config for TestRuntime {
+//!     ..
+//!     // we only care about these two for now. The rest can be mock. The block number type of
+//!     // kusama is u32.
+//!     type BlockNumber = u32;
+//!     type Header = Header;
+//!     ..
+//! }
+//!
+//! #[test]
+//! fn test_runtime_works() {
+//!     let hash: Hash =
+//!         hex!["f9a4ce984129569f63edc01b1c13374779f9384f1befd39931ffdcc83acf63a7"].into();
+//!     let parent: Hash =
+//!         hex!["540922e96a8fcaf945ed23c6f09c3e189bd88504ec945cc2171deaebeaf2f37e"].into();
+//!     Builder::new()
+//!         .at(hash)
+//!         .module("System")
+//!         .build()
+//!         .execute_with(|| {
+//!             assert_eq!(
+//!                 // note: the hash corresponds to 3098546. We can check only the parent.
+//!                 // https://polkascan.io/kusama/block/3098546
+//!                 <frame_system::Module<Runtime>>::block_hash(3098545u32),
+//!                 parent,
+//!             )
+//!         });
+//! }
+//! ```
+//!
+//! Or with the real kusama runtime.
+//!
+//! ```ignore
+//! use remote_externalities::Builder;
+//! use kusama_runtime::Runtime;
+//!
+//! #[test]
+//! fn test_runtime_works() {
+//!     let hash: Hash =
+//!         hex!["f9a4ce984129569f63edc01b1c13374779f9384f1befd39931ffdcc83acf63a7"].into();
+//!     Builder::new()
+//!         .at(hash)
+//!         .module("Staking")
+//!         .build()
+//!         .execute_with(|| assert_eq!(<pallet_staking::Module<Runtime>>::validator_count(), 400));
+//! }
+//! ```
+
+use std::{
+	fs,
+	path::{Path, PathBuf},
+};
+use std::fmt::{Debug, Formatter, Result as FmtResult};
+use log::*;
+use sp_core::{hashing::twox_128};
+pub use sp_io::TestExternalities;
+use sp_core::storage::{StorageKey, StorageData};
+use futures::future::Future;
+
+type KeyPair = (StorageKey, StorageData);
+type Number = u32;
+type Hash = sp_core::H256;
+// TODO: make these two generic.
```

Comment:
I would be happy to do this in a future PR. Given that we want to use this initially for known chains, I think it is an overkill for now.